src/Params/WaveTable.h

```diff
@@ -0,0 +1,659 @@
+    size_t r = 0;
+    //! write position, marks the element that will be written to next (may be
+    //! reserved for write already if w_delayed < w). can write until w.
+    size_t w_delayed = 0;
+    //! write position, marks the element that will be reserved for a write
+    //! next (but maybe not yet written). can write until r-1.
+    size_t w = 0;
+
+    // const (except of swapping)
+    size_t m_size;         //!< size
+    size_t m_size_mask;    //!< size - 1
+    size_t m_size_x2;      //!< size * 2
+    size_t m_size_x2_mask; //!< size * 2 - 1
+};
+
+/*
+ * Wavetable
+ */
+
+tensor_size_t findBestIndex(const Tensor<1, wavetable_types::float32>& freqs, float freq);
+
+using Shape1 = Shape<1>;
+using Shape2 = Shape<2>;
+using Shape3 = Shape<3>;
+
+//! Tensor3 that keeps track of its subtensors, using a Tensor of
+//! AbstractRingbuffer
+class Tensor3ForWaveTable : public Tensor<3, wavetable_types::float32>
+{
+    using base_type = Tensor<3, wavetable_types::float32>;
+
+public:
+    Tensor3ForWaveTable(const Shape<3>& shape)
+        : base_type(shape), ringbuffers(shape.dim[1])
+    {
+        for(AbstractRingbuffer& rb : ringbuffers)
+            rb.resize(shape.dim[0]);
+    }
+
+    void swapWith(Tensor3ForWaveTable& other)
+    {
+        base_type::swapWith(other);
+        ringbuffers.swapWith(other.ringbuffers);
+    }
+
+    Tensor<1, AbstractRingbuffer> ringbuffers;
+};
+
+/**
+    All pre-computed data that ADnote needs for generating a note
+    @note This class will only ever reside in the RT thread. Non-RT threads
+          may generate the required tensors, but only send them via bToU.
+          At no time, non-RT threads access content directly.
+*/
+class WaveTable
+{
+public:
+    using float32 = wavetable_types::float32;
+    using IntOrFloat = wavetable_types::IntOrFloat;
+    using WtMode = wavetable_types::WtMode;
+
+    // pure guesses for what sounds good:
+    constexpr const static tensor_size_t num_freqs = 10;
+    constexpr const static tensor_size_t num_semantics = 32; // (in non-WT mode)
+    // number of steps on basefunc knob (the steps are integral)
+    constexpr const static float wt_127f = 127.f;
+    // in WT mod, we have steps of 0.25. This variable adds 3 unused waves padding.
+    constexpr const static tensor_size_t max_semantics_ever = 128 * 4;
+    constexpr const static tensor_size_t num_semantics_wtmod = max_semantics_ever;
+
+    static_assert (num_semantics <= max_semantics_ever, "max_semantics_ever too low");
+    static_assert (num_semantics_wtmod <= max_semantics_ever, "max_semantics_ever too low");
+
+private:
+    //! Semantics for which the waves are built.
+    //! E.g. oscil params or random seed (e.g. 0...127).
+    Tensor1<IntOrFloat> semantics;
+    //! Frequencies for which the waves are built.
+    Tensor1<float32> freqs;
+    //! The waves.
+    //! 1st dim: semantics, 2nd: freqs, 3rd: time.
+    Tensor3ForWaveTable data;
```

Comment:
ok, I'm currently extremely confused why indexing over semantics/freq yields ringbuffers as opposed to yielding float32 arrays of some shared size. Perhaps there was some considerable refactoring into the adnote code which converts from float32 buffers into output streams. I'm continuing to read onwards

Reply:
I also got confused a bit for a few minutes. Tensor3ForWaveTable contains two things:

* `Tensor<3, float>` by inheritance
* `Tensor<1, AbstractRingbuffer>` as a member

The `Tensor<3, float>` indeed contains the float32 that you are looking for. The important thing about the ringbuffers is that they are all abstract. They do contain valid indices, but no buffers - the indices just refer to the `Tensor<3, float>` buffers.

So it is rather simple - no feature like converting float32 buffers into output streams.

Reply:
ok, then the Tensor<3, float> makes sense, but I'm still a bit confused with the ring buffers exact role, though I recall they were mentioned in the overall design doc so I guess I should pivot back to there when I'm through Adnote.cpp

Reply:
The ringbuffers have only 1 purpose. In case of non-WT mode, and if random seeds are in use, then we both consume waves and generate new ones. So these "semantics" have a producer-consumer situation which is realized with a ringbuffer (for each frequency separately, since consuming a wave at a higher frequency does not mean that you need to regenerate the wave at a lower frequency - this is why in "Send the results to RT" in MiddleWare.cpp, sometimes only 1-dimensional floats are requested).